auth0/src/main/java/com/auth0/android/provider/SignatureVerifier.java

```diff
@@ -0,0 +1,62 @@
+package com.auth0.android.provider;
+
+import android.support.annotation.NonNull;
+
+import com.auth0.android.authentication.AuthenticationAPIClient;
+import com.auth0.android.authentication.AuthenticationException;
+import com.auth0.android.callback.AuthenticationCallback;
+import com.auth0.android.callback.BaseCallback;
+import com.auth0.android.jwt.JWT;
+
+import java.security.InvalidKeyException;
+import java.security.PublicKey;
+import java.util.Map;
+
+/**
+ * Abstract class meant to verify tokens signed with HS256 and RS256 signatures.
+ */
+abstract class SignatureVerifier {
+
+    /**
+     * Verifies that the given token's signature is valid, deeming the payload inside it authentic
+     *
+     * @param token the ID token to have its signature validated
+     * @throws TokenValidationException if the signature is not valid
+     */
+    abstract void verifySignature(@NonNull JWT token) throws TokenValidationException;
+
+    /**
+     * Validates the algorithm of the given token is supported and creates a new instance of a SignatureVerifier
+     *
+     * @param token     the ID token to create a signature verifier for
+     * @param apiClient api client instance to fetch the JWKS keys, if necessary
+     * @param callback  the callback to receive the result in
+     */
+    static void forToken(@NonNull JWT token, @NonNull AuthenticationAPIClient apiClient, @NonNull final BaseCallback<SignatureVerifier, TokenValidationException> callback) {
+        String algorithmName = token.getHeader().get("alg");
+        if ("RS256".equals(algorithmName)) {
+            final String keyId = token.getHeader().get("kid");
```

Comment:
What if `kid` is empty?

Reply:
then `keyId` will end up with a `null` value, and when used below there will be no results. Thus, creating an instance of the `AsymmetricSignatureVerifier` will throw. Covered [here](https://github.com/auth0/Auth0.Android/pull/265/files#diff-c422ae1e5c4358e1866a66f51b9440b5R1696)

Reply:
Added an explicit test case for when the claim `keyId` is not set on the header.